File: src/main/java/utilities/GoalParser.java
package utilities;

import entities.*;
import interface_adapters.Obtainable;


import com.google.gson.*;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.*;

public class GoalParser {

    private static final Set<String> LANTERN_REQUIRED = new HashSet<>(Arrays.asList("No Eyes", "Peaks Access"));

    public static ArrayList<Obtainable> parseGoals(InputStream inputStream, SkipSettings settings) {
        try (Reader reader = new InputStreamReader(inputStream, StandardCharsets.UTF_8);) {
            JsonParser parser = new JsonParser();
            JsonObject object = parser.parse(reader).getAsJsonObject();
            JsonArray jsonGoals = object.getAsJsonArray("goals");

            Map<String, Obtainable> byName = new HashMap<>();
            ArrayList<Obtainable> result = new ArrayList<>();

            // First pass: create empty objects
            for (JsonElement element : jsonGoals) {
                JsonObject obj = element.getAsJsonObject();

                String name = obj.get("name").getAsString();
                String type = obj.get("type").getAsString();

                Obtainable goal;

                if (type.equals("CollectionGoal")) {
                    int count = 0;
                    if (obj.has("count")) {
                        count = obj.get("count").getAsInt();
                    }
                    goal = new CollectionGoal(
                            name,
                            new ArrayList<>(),
                            count
                    );
                } else if (type.equals("Objective")) {
                    goal = new Objective(
                            name,
                            new ArrayList<>()
                    );
                } else if (type.equals("AchievementGoal")){
                    goal = new AchievementGoal(
                            name,
                            new ArrayList<>()
                    );
                } else if (type.equals("MilestoneGoal")) {
                    int count = 0;
                    if (obj.has("count")) {
                        count = obj.get("count").getAsInt();
                    }

                    String objective = null;
                    if (obj.has("objective")) {
                        objective = String.valueOf(obj.get("objective"));
                    }
                    goal = new MilestoneGoal(name, objective, count);
                } else {
                    goal = null;
                }

                byName.put(name, goal);
                result.add(goal);
            }

            // Second pass: fill dependencies/options
            for (JsonElement element : jsonGoals) {
                JsonObject obj = element.getAsJsonObject();
                String name = obj.get("name").getAsString();
                Obtainable goal = byName.get(name);

                if (!obj.has("options")) {
                    continue;
                }

                // CollectionGoal special handling
                if (goal instanceof CollectionGoal) {
                    for (JsonElement optionElement : obj.getAsJsonArray("options")) {
                        String dependencyName = optionElement.getAsString();
                        Obtainable dependency = byName.get(dependencyName);
                        if (dependency == null) {
                            throw new RuntimeException(
                                    "Unknown collection dependency: " + dependencyName
                            );
                        }
                        ((CollectionGoal) goal).getCollectionItems().add(dependency);
                    }
                    continue;
                }

                // MilestoneGoal special handling
                if (goal instanceof MilestoneGoal) {
                    continue;
                }

                // Normal ObtainOption handling
                ArrayList<ObtainOption> options = goal.getDependencies();

                for (JsonElement optionElement : obj.getAsJsonArray("options")) {
                    JsonObject optionObj = optionElement.getAsJsonObject();
                    HashSet<Obtainable> dependencies = new HashSet<>();

                    if (optionObj.has("notes")) {
                        String note = optionObj.get("notes").getAsString();
                        if (!settings.getSettings().contains(note)) {
                            continue; //dont process
                        }
                    }

                    if (optionObj.has("dependencies")) {
                        for (JsonElement depElement : optionObj.getAsJsonArray("dependencies")) {
                            String depName = depElement.getAsString();
                            Obtainable dependency = byName.get(depName);
                            if (dependency == null) {
                                throw new RuntimeException(
                                        "Unknown dependency: " + depName
                                );
                            }
                            dependencies.add(dependency);
                        }
                    }
                    PlayerStateEffect effect = parseEffect(optionObj.getAsJsonObject("effect"));
                    options.add(new ObtainOption(dependencies, effect));
                }
            }
            if (settings.hasDarkrooms()) {
                System.out.println("darkrooms on");
                removeLantern(GoalUtility.getGoalByName(result, "Lumafly Lantern"),
                                                    result);
            }
            return result;

        } catch (Exception e) {
            throw new RuntimeException("Failed to parse goals JSON", e);
        }
    }

    private static void removeLantern(Obtainable lumaflyLantern, ArrayList<Obtainable> result) {
        for (Obtainable l : result) {
            if (LANTERN_REQUIRED.contains(l.getName())) {
                continue;
            }

            for (ObtainOption o : l.getDependencies()) {
                o.getDependencies().remove(lumaflyLantern);
            }
        }
    }

    private static PlayerStateEffect parseEffect(JsonObject effectObj) {
        if (effectObj == null) {
            return new PlayerStateEffect();
        }

        int grubs = getInt(effectObj, "grubs_saved");
        int geo = getInt(effectObj, "geo");
        int essence = getInt(effectObj, "essence");
        int tolls = getInt(effectObj, "tolls");

        return new PlayerStateEffect(grubs, Math.abs(geo), essence, tolls);
    }

    private static int getInt(JsonObject obj, String key) {
        if (obj.has(key) && !obj.get(key).isJsonNull()) {
            return obj.get(key).getAsInt();
        }
        return 0;
    }
}

File: gh-pages/src/js/goal_parser.ts
"use strict";

import { AchievementGoal, CollectionObtainable, MilestoneGoal, Objective, Obtainable, PlayerStateEffect } from "./entities";
import { SkipSettings } from "./settings";

type Option = {
  dependencies: Array<string>;
  effect?: {
    geo?: number;
    simple_keys?: number;
    pale_ore?: number;
    grubs_saved?: number;
    tolls?: number;
    lifeblood_masks?: number;
  };
  notes?: string;
};

type JSONGoal = {
  name: string;
  type: string;
  options?: Array<Option | string>;
};

export function parseGoals(goals: Array<JSONGoal>, settings: SkipSettings): Array<Obtainable> {
  const byName: Record<string, Obtainable> = {};
  const result: Array<Obtainable> = Array();

  const getOrCreate = (name: string) => {
    if (!(name in byName)) {
      byName[name] = { name, options: [], goalKind: "unknown" }
    }
    return byName[name]
  }


  // First pass: create empty objects
  for (const element of goals) {
    let newGoal: Obtainable = getOrCreate(element.name);
    const type: string = element.type;

    if (type === "CollectionGoal") {
      if ("count" in element) {
        // @ts-expect-error
        let collectionGoal: CollectionObtainable = newGoal;
        collectionGoal.goalKind = "CollectionObtainable";
        collectionGoal.count = element.count as number;
        collectionGoal.collectionItems = (element.options as string[]).map(getOrCreate);
        newGoal = collectionGoal;
      }
      else {
        alert(`error parsing goal ${element.name} of kind CollectionGoal. Everything will break now`);
      }
    } else if (type === "MilestoneGoal") {
      if ("amount" in element && "objective" in element) {
        // @ts-expect-error
        let milestoneGoal: MilestoneGoal = newGoal;
        milestoneGoal.goalKind = "MilestoneGoal";
        milestoneGoal.milestoneType = element.objective as string;
        milestoneGoal.milestone = element.amount as number;
        newGoal = milestoneGoal;
      }
      else {
        alert(`error parsing goal ${element.name} of kind MilestoneGoal. Everything will break now`);
      }
    } else if (type === "Objective") {
      (newGoal as Objective).goalKind = "Objective";
    } else if (type === "AchievementGoal") {
      (newGoal as AchievementGoal).goalKind = "AchievementGoal";
    } else {
      alert(`Unknown type of goal ${type} for goal ${element.name}. Everything will break now`);
    }
    // fill dependencies/options for Objective and AchievementGoal
    if ((newGoal.goalKind === "Objective" || newGoal.goalKind === "AchievementGoal")
      && "options" in element) {
      const options = newGoal.options;

      for (const optionObj of element.options) {
        const option = (optionObj as Option);
        const dependencies = [];
        if (!!option.notes && !settings.settings.includes(option.notes)) {
          continue; //dont process
        }

        if ("dependencies" in option) {
          for (const depElement of option.dependencies) {
            const depName = depElement;
            const dependency = getOrCreate(depName);
            dependencies.push(dependency);
          }
        }
        const effect: PlayerStateEffect = parseEffect(option.effect);
        options.push({ dependencies, effect });
      }
    }
    result.push(newGoal);
  }

  if (settings.darkrooms) {
    console.log("darkrooms on");
    removeLantern(result);
  }
  return result;
}

const LANTERN_REQUIRED = ["No Eyes", "Peaks Access"];

function removeLantern(result: Array<Obtainable>): Array<Obtainable> {
  for (const l of result) {
    if (LANTERN_REQUIRED.includes(l.name)) {
      continue;
    }

    for (const option of l.options) {
      // In-place removal of dependency
      const idx = option.dependencies.findIndex(dep => dep.name === "Lumafly Lantern");
      if (idx > -1) {
        option.dependencies.splice(idx, 1);
      }
    }
  }
  return result
}

function parseEffect(effectInput: Option["effect"] | null | undefined): PlayerStateEffect {
  const effect = {
    grubs_collected: 0,
    geo_spent: 0,
    essence_collected: 0,
    tolls_collected: 0
  };
  if (!!effectInput) {
    effect.grubs_collected = effectInput.grubs_saved ?? 0;
    effect.geo_spent = Math.abs(effectInput.geo ?? 0);
    // effect.essence_collected = effectInput.essence ?? 0;
    effect.tolls_collected = effectInput.tolls ?? 0;
  }
  return effect;
}
